pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,109 @@
+
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/controller"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/manager"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/winsvc"
+	"github.com/openshift/windows-machine-config-operator/pkg/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/pkg/servicescm"
+)
+
+// Execute removes an instance from the cluster. If the parameter flag is true, the node object is not deleted.
+// If we are able to get the services ConfigMap tied to the desired version, all services defined in it are cleaned up.
+// TODO: Best effort cleanup if node object, desired version annotation, or services ConfigMap are not attainable
+func Execute(c client.Client, addrs []net.Addr, svcMgr manager.Manager, preserveNode bool) error {
+	ctx := context.TODO()
+	node, err := controller.GetAssociatedNode(c, addrs)
+	if err != nil {
+		klog.Exitf("could not node object associated with this instance: %s", err.Error())
+	}
+	klog.Infof("cleaning up managed services on node %s", node.Name)
+
+	desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+	if !present {
+		return errors.Errorf("node missing desired version annotation")
+	}
+	// Fetch the CM of the desired version
+	cm := &core.ConfigMap{}
+	err = c.Get(ctx,
+		client.ObjectKey{Namespace: controller.WmcoNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	if err != nil {
+		return err
+	}
+	return cleanCMServices(c, ctx, svcMgr, cm, node, preserveNode)
+}
+
+// cleanCMServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the given flag is false
+func cleanCMServices(c client.Client, ctx context.Context, svcMgr manager.Manager, cm *core.ConfigMap, node *core.Node,
+	preserveNode bool) error {
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	existingSvcs, err := svcMgr.GetExistingServices()
+	if err != nil {
+		return err
+	}
+
+	klog.Infof("removing services defined in ConfigMap %s", cm.Name)
+	// Build up any failures
+	failedRemovals := []string{}
+	// The services are ordered by increasing priority already, so stop them in reverse order to avoid dependency issues
+	for i := len(cmData.Services) - 1; i >= 0; i-- {
+		service := cmData.Services[i]
+		if _, ok := existingSvcs[service.Name]; !ok {
+			// Nothing to do if it already does not exist
+			continue
+		}
+		// Open the service to get a service handle
+		winSvcObj, err := svcMgr.OpenService(service.Name)
+		if err != nil {
+			failedRemovals = append(failedRemovals, err.Error())
+		}
+		// Ensure service is stopped before deleting
+		if err := winsvc.EnsureServiceState(winSvcObj, svc.Stopped); err != nil {
+			failedRemovals = append(failedRemovals, err.Error())
+		}
+		klog.Infof("removing service %s", service.Name)
+		if err := svcMgr.DeleteService(service.Name); err != nil {
```

Comment:
I assume you meant  "Can you _delete_ a service that you were not able to open/stop?" -- the answer is no. 
Note that the happy path (where we can get the desired services CM for the node) is not best effort. It should always be able to find, stop, and delete the services listed in the ConfigMap, otherwise that is an error worthy event.

Reply:
>Can you delete a service that you were not able to open/stop?" -- the answer is no.

Then you should be logging and continuing not moving on to deleting on error when opening or stopping.

>Note that the happy path (where we can get the desired services CM for the node) is not best effort. It should always be able to find, stop, and delete the services listed in the ConfigMap, otherwise that is an error worthy event.

I don't agree with that. It always should be best effort, while you collect and display all the errors you hit. Stopping at the first error should not be the option.

Reply:
> not moving on to deleting on error when opening or stopping
> Stopping at the first error should not be the option.

Agreed on both